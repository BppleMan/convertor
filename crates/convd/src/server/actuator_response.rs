use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActuatorResponse<T> {
    pub status: u16,
    pub message: String,
    pub data: Option<T>,
}

impl<T> Default for ActuatorResponse<T>
where
    T: Serialize + DeserializeOwned,
{
    fn default() -> Self {
        Self {
            status: ActuatorResponseStatus::OK,
            message: "OK".to_string(),
            data: None,
        }
    }
}

impl<T> ActuatorResponse<T>
where
    T: Serialize + DeserializeOwned,
{
    pub fn ok() -> ActuatorResponse<T> {
        Self {
            status: 0,
            message: "OK".to_string(),
            data: None,
        }
    }

    pub fn with_data(mut self, data: T) -> Self {
        self.data = Some(data);
        self
    }

    pub fn ok_data(data: T) -> Self {
        Self {
            status: 200,
            message: "OK".to_string(),
            data: Some(data),
        }
    }
}

#[derive(Default, Serialize, Deserialize)]
pub enum ActuatorResponseStatus {
    #[default]
    OK = 0,
}
