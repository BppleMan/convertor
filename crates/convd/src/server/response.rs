mod api_error;
mod api_response;
mod app_error;
mod request_snapshot;

pub use api_error::*;
pub use api_response::*;
pub use app_error::*;
pub use request_snapshot::*;
