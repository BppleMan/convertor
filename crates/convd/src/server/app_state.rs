use crate::server::service::{ClashService, SurgeService};
use convertor::config::ConvertorConfig;
use convertor::config::provider_config::Provider;
use convertor::provider_api::ProviderApi;
use std::collections::HashMap;
use std::sync::Arc;

#[derive(Clone)]
pub struct AppState {
    pub config: Arc<ConvertorConfig>,
    pub api_map: HashMap<Provider, ProviderApi>,
    pub surge_service: SurgeService,
    pub clash_service: ClashService,
}

impl AppState {
    pub fn new(config: ConvertorConfig, api_map: HashMap<Provider, ProviderApi>) -> Self {
        let config = Arc::new(config);
        let surge_service = SurgeService::new(config.clone());
        let clash_service = ClashService::new(config.clone());
        Self {
            config,
            api_map,
            surge_service,
            clash_service,
        }
    }
}
