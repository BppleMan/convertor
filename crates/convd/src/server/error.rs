use axum::http::StatusCode;
use axum::http::header::ToStrError;
use axum::response::{IntoResponse, Response};
use convertor::common::config::proxy_client_config::ProxyClient;
use convertor::url::url_error::{QueryError, UrlBuilderError};
use std::sync::Arc;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum AppError {
    #[error("获取原始非转换配置失败, 遇到错误的客户端: {0}")]
    RawProfileUnsupportedClient(ProxyClient),

    #[error("没有找到对应的订阅提供者")]
    NoSubProvider,

    // #[error("没有找到对应的代理客户端")]
    // NoProxyClient,
    #[error(transparent)]
    ConvertorUrl(#[from] UrlBuilderError),

    #[error("Unauthorized: {0}")]
    Unauthorized(String),

    #[error(transparent)]
    QueryError(#[from] QueryError),

    #[error(transparent)]
    Eyre(#[from] color_eyre::Report),

    #[error(transparent)]
    ToStr(#[from] ToStrError),

    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),

    #[error(transparent)]
    CacheError(#[from] Arc<AppError>),
    // #[error("读写锁异常: {0}")]
    // RwLockError(String),
}

impl IntoResponse for AppError {
    fn into_response(self) -> Response {
        let message = format!("{self:?}");
        let message = console::strip_ansi_codes(&message).to_string();
        (StatusCode::INTERNAL_SERVER_ERROR, message).into_response()
    }
}
