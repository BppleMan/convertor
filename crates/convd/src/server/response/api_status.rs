use crate::server::response::ApiResponse;
use axum::http::{HeaderValue, StatusCode, header};
use axum::response::{IntoResponse, Response};
use serde::{Deserialize, Serialize};
use std::borrow::Cow;
use tokio_util::bytes::{BufMut, BytesMut};

#[derive(Default, Clone, Serialize, Deserialize)]
pub struct ApiStatus {
    pub code: i64,
    pub message: Cow<'static, str>,
}

impl ApiStatus {
    pub const OK: Self = ApiStatus {
        code: 0,
        message: Cow::Borrowed("ok"),
    };

    pub const ERROR: Self = ApiStatus {
        code: 1,
        message: Cow::Borrowed("error"),
    };
}

impl ApiStatus {
    pub fn with_code(mut self, code: i64) -> Self {
        self.code = code;
        self
    }

    pub fn with_message(mut self, message: impl AsRef<str>) -> Self {
        self.message = Cow::Owned(message.as_ref().to_string());
        self
    }
}

impl<T> IntoResponse for ApiResponse<T>
where
    T: Serialize,
{
    fn into_response(self) -> Response {
        // Extracted into separate fn so it's only compiled once for all T.
        fn make_response(buf: BytesMut, ser_result: serde_json::Result<()>) -> Response {
            match ser_result {
                Ok(()) => (
                    [(
                        header::CONTENT_TYPE,
                        HeaderValue::from_static(mime::APPLICATION_JSON.as_ref()),
                    )],
                    buf.freeze(),
                )
                    .into_response(),
                Err(err) => (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    [(
                        header::CONTENT_TYPE,
                        HeaderValue::from_static(mime::TEXT_PLAIN_UTF_8.as_ref()),
                    )],
                    err.to_string(),
                )
                    .into_response(),
            }
        }

        // Use a small initial capacity of 128 bytes like serde_json::to_vec
        // https://docs.rs/serde_json/1.0.82/src/serde_json/ser.rs.html#2189
        let mut buf = BytesMut::with_capacity(128).writer();
        let res = serde_json::to_writer(&mut buf, &self);
        make_response(buf.into_inner(), res)
    }
}
