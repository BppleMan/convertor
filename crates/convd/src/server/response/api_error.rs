use axum::http::StatusCode;
use axum::http::header::ToStrError;
use axum::response::{IntoResponse, Response};
use convertor::config::proxy_client::ProxyClient;
use convertor::error::{ProviderError, QueryError, UrlBuilderError};
use redis::RedisError;
use std::sync::Arc;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ApiError {
    #[error("获取原始非转换配置失败, 遇到错误的客户端: {0}")]
    RawProfileUnsupportedClient(ProxyClient),

    #[error("没有找到对应的订阅提供者")]
    NoSubProvider,

    #[error(transparent)]
    UrlBuilderError(#[from] UrlBuilderError),

    #[error("Unauthorized: {0}")]
    Unauthorized(String),

    #[error(transparent)]
    QueryError(#[from] QueryError),

    #[error(transparent)]
    ProviderError(#[from] ProviderError),

    #[error(transparent)]
    ToStr(#[from] ToStrError),

    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),

    #[error(transparent)]
    CacheError(#[from] Arc<ApiError>),

    #[error("Redis 错误: {0:?}")]
    RedisError(#[from] RedisError),

    #[error(transparent)]
    Unknown(#[from] color_eyre::Report),
}

impl IntoResponse for ApiError {
    fn into_response(self) -> Response {
        let message = format!("{self}");
        (StatusCode::INTERNAL_SERVER_ERROR, message).into_response()
    }
}
