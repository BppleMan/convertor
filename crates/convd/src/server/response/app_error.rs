use crate::server::response::{ApiResponse, ApiStatus};
use axum::http::header::ToStrError;
use convertor::config::proxy_client::ProxyClient;
use convertor::error::{ParseError, ProviderError, QueryError, RenderError, UrlBuilderError};
use redis::RedisError;
use std::sync::Arc;
use thiserror::Error;

macro_rules! define_error {
    (
        $(#[$enum_attr:meta])*
        $vis:vis enum $name:ident {
            $(
                $(#[$var_attr:meta])*
                $var:ident
                $( ( $( $(#[$fattr:meta])* $fty:ty ),+ $(,)? ) )?
                $( { $( $(#[$field_attr:meta])* $fname:ident : $ftyt:ty ),+ $(,)? } )?
                ; $vi:expr
                $(,)?
            )*
        }
    ) => {
        $(#[$enum_attr])*
        $vis enum $name {
            $(
                $(#[$var_attr])*
                $var
                $( ( $( $(#[$fattr])* $fty ),+ ) )?
                $( { $( $(#[$field_attr])* $fname : $ftyt ),+ } )?
            ),*
        }

        impl From<$name> for ApiResponse<()> {
            fn from(value: $name) -> Self {
                ApiResponse::error_with_status(value.into())
            }
        }

        impl From<AppError> for ApiStatus {
            fn from(value: AppError) -> Self {
                match value {
                    $($name::$var(e) => Self::new($vi, e),)*
                }
            }
        }

    };
}

define_error! {
    /// 用于表示应用程序中的各种错误类型。
    #[allow(clippy::large_enum_variant)]
    #[derive(Debug, Error)]
    pub enum AppError {
        #[error("获取订阅商原始配置失败, 遇到错误的客户端: {0}")]
        RawProfileUnsupportedClient(ProxyClient) ; 1001,

        #[error("Unauthorized: {0}")]
        Unauthorized(String) ; 1002,

        #[error(transparent)]
        UrlBuilderError(#[from] UrlBuilderError) ; 1003,

        #[error(transparent)]
        QueryError(#[from] QueryError) ; 1004,

        #[error(transparent)]
        ProviderError(#[from] ProviderError) ; 1005,

        #[error(transparent)]
        ParseError(#[from] ParseError) ; 1006,

        #[error(transparent)]
        RenderError(#[from] RenderError) ; 1007,

        #[error(transparent)]
        ToStr(#[from] ToStrError) ; 1008,

        #[error(transparent)]
        Utf8Error(#[from] std::str::Utf8Error) ; 1009,

        #[error(transparent)]
        CacheError(#[from] Arc<AppError>) ; 1010,

        #[error("Redis 错误: {0:?}")]
        RedisError(#[from] RedisError) ; 1011,

        #[error(transparent)]
        JsonError(#[from] serde_json::Error) ; 1012,
    }
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, Error)]
pub enum SnapshotError {
    #[error("获取订阅商原始配置失败, 遇到错误的客户端: {0}")]
    RawProfileUnsupportedClient(ProxyClient),

    #[error("Unauthorized: {0}")]
    Unauthorized(String),

    #[error(transparent)]
    UrlBuilderError(#[from] UrlBuilderError),

    #[error(transparent)]
    QueryError(#[from] QueryError),

    #[error(transparent)]
    ProviderError(#[from] ProviderError),

    #[error(transparent)]
    ParseError(#[from] ParseError),

    #[error(transparent)]
    RenderError(#[from] RenderError),

    #[error(transparent)]
    ToStr(#[from] ToStrError),

    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),

    #[error(transparent)]
    CacheError(#[from] Arc<SnapshotError>),

    #[error("Redis 错误: {0:?}")]
    RedisError(#[from] RedisError),

    #[error(transparent)]
    JsonError(#[from] serde_json::Error),
}

// impl From<AppError> for ApiResponse<()> {
//     fn from(value: AppError) -> Self {
//         ApiResponse::error_with_status(value.into())
//     }
// }

// impl From<AppError> for ApiStatus {
//     fn from(value: AppError) -> Self {
//         match value {
//             AppError::RawProfileUnsupportedClient(e) => Self::new(1001, e),
//             AppError::UrlBuilderError(e) => Self::new(1002, e),
//             AppError::Unauthorized(e) => Self::new(1003, e),
//             AppError::QueryError(e) => Self::new(1005, e),
//             AppError::ProviderError(e) => Self::new(1006, e),
//             AppError::ParseError(e) => Self::new(9, e),
//             AppError::RenderError(e) => Self::new(1004, e),
//             AppError::ToStr(e) => Self::new(1007, e),
//             AppError::Utf8Error(e) => Self::new(1008, e),
//             AppError::CacheError(e) => Self::new(1009, e),
//             AppError::RedisError(e) => Self::new(10010, e),
//             AppError::JsonError(e) => Self::new(1011, e),
//         }
//     }
// }
