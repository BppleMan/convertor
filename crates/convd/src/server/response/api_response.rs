use crate::server::response::api_error::ApiError;
use crate::server::response::{AppError, RequestSnapshot};
use axum::http::{HeaderValue, StatusCode, header};
use axum::response::{IntoResponse, Response};
use serde::{Deserialize, Serialize};
use std::borrow::Cow;
use std::fmt::Display;
use tokio_util::bytes::{BufMut, BytesMut};

#[derive(Default, Clone, Serialize, Deserialize)]
pub struct ApiResponse<T>
where
    T: serde::Serialize,
{
    pub status: String,
    pub messages: Vec<Cow<'static, str>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request: Option<RequestSnapshot>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<T>,
}

impl<T> ApiResponse<T>
where
    T: serde::Serialize,
{
    pub fn ok(data: T) -> Self {
        Self {
            status: "ok".to_string(),
            messages: vec![],
            request: None,
            data: Some(data),
        }
    }

    pub fn with_message(mut self, message: impl Display) -> Self {
        self.messages = vec![Cow::Owned(message.to_string())];
        self
    }

    pub fn with_request(mut self, request: RequestSnapshot) -> Self {
        self.request = Some(request);
        self
    }
}

impl ApiResponse<()> {
    pub fn from_error(status: impl Display, error: impl core::error::Error) -> Self {
        let status = status.to_string();
        let mut messages = vec![Cow::Owned(error.to_string())];
        let mut source = error.source();
        while let Some(src) = source {
            messages.push(Cow::Owned(src.to_string()));
            source = src.source();
        }
        Self {
            status,
            messages,
            request: None,
            data: None::<()>,
        }
    }
}

impl<T> IntoResponse for ApiResponse<T>
where
    T: Serialize,
{
    fn into_response(self) -> Response {
        let mut buf = BytesMut::with_capacity(256).writer();
        match serde_json::to_writer(&mut buf, &self) {
            Ok(()) => (
                [(
                    header::CONTENT_TYPE,
                    HeaderValue::from_static(mime::APPLICATION_JSON.as_ref()),
                )],
                buf.into_inner().freeze(),
            )
                .into_response(),
            Err(err) => {
                let api_error = ApiError {
                    status: StatusCode::INTERNAL_SERVER_ERROR,
                    error: AppError::JsonError(err),
                    request: None,
                };
                api_error.into_response()
            }
        }
    }
}
