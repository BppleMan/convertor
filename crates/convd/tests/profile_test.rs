#[path = "./server.rs"]
mod server;

use crate::server::{ServerContext, start_server};
use axum::body::Body;
use axum::extract::Request;
use axum::response::Response;
use convertor::config::client_config::ProxyClient;
use convertor::config::provider_config::Provider;
use convertor::testkit::init_test;
use http_body_util::BodyExt;
use tower::ServiceExt;

async fn profile(
    server_context: &ServerContext,
    client: ProxyClient,
    provider: Provider,
) -> color_eyre::Result<String> {
    let ServerContext { app, app_state, .. } = server_context;
    let url_builder = app_state.config.create_url_builder(client, provider)?;

    let profile_url = url_builder.build_profile_url()?;
    let request = Request::builder()
        .uri(profile_url.to_string())
        .method("GET")
        .header("host", "127.0.0.1")
        .header("user-agent", concat!("convertor/", env!("CARGO_PKG_VERSION")))
        .body(Body::empty())?;
    let response: Response = app.clone().oneshot(request).await?;

    let actual = String::from_utf8_lossy(&response.into_body().collect().await?.to_bytes())
        .to_string()
        .replace(
            &url_builder.enc_sub_url,
            "http://127.0.0.1:8080/subscription?token=bppleman",
        );
    Ok(actual)
}

#[tokio::test]
async fn test_profile_surge_boslife() -> color_eyre::Result<()> {
    init_test();
    let server_context = start_server().await?;
    let actual = profile(&server_context, ProxyClient::Surge, Provider::BosLife).await?;
    insta::assert_snapshot!(actual);
    Ok(())
}

#[tokio::test]
async fn test_profile_clash_boslife() -> color_eyre::Result<()> {
    init_test();
    let server_context = start_server().await?;
    let actual = profile(&server_context, ProxyClient::Clash, Provider::BosLife).await?;
    insta::assert_snapshot!(actual);
    Ok(())
}
