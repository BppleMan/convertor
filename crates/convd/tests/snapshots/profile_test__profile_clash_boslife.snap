---
source: crates/convd/tests/profile_test.rs
expression: actual
---
port: 7890
socks-port: 7891
redir-port: 7892
allow-lan: true
mode: rule
log-level: info
external-controller: 0.0.0.0:9091
external-ui: ui
secret: ""

proxies:
    - { name: "订阅提供商信息", type: "ss", server: "bppleman.com", port: 24000, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇭🇰 香港 01", type: "ss", server: "bppleman.com", port: 24000, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇭🇰 香港 02", type: "ss", server: "bppleman.com", port: 24001, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇭🇰 香港 03", type: "ss", server: "bppleman.com", port: 24002, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇭🇰 香港 04", type: "ss", server: "bppleman.com", port: 24003, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇭🇰 香港 05 - YouTube免广", type: "ss", server: "bppleman.com", port: 24004, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🏳️‍🌈 台湾 01", type: "ss", server: "bppleman.com", port: 23010, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🏳️‍🌈 台湾 02", type: "ss", server: "bppleman.com", port: 23011, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🏳️‍🌈 台湾 03", type: "ss", server: "bppleman.com", port: 23012, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🏳️‍🌈 台湾 04", type: "ss", server: "bppleman.com", port: 23013, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🏳️‍🌈 台湾 05", type: "ss", server: "bppleman.com", port: 23014, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇸🇬 新加坡 01", type: "ss", server: "bppleman.com", port: 23001, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇸🇬 新加坡 02", type: "ss", server: "bppleman.com", port: 23002, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇸🇬 新加坡 03", type: "ss", server: "bppleman.com", port: 23003, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇸🇬 新加坡 04", type: "ss", server: "bppleman.com", port: 23004, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇸🇬 新加坡 05", type: "ss", server: "bppleman.com", port: 23005, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇯🇵 日本 01", type: "ss", server: "bppleman.com", port: 23020, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇯🇵 日本 02", type: "ss", server: "bppleman.com", port: 23021, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇯🇵 日本 03", type: "ss", server: "bppleman.com", port: 23022, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇯🇵 日本 04", type: "ss", server: "bppleman.com", port: 23023, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇯🇵 日本 05", type: "ss", server: "bppleman.com", port: 23024, password: "bppleman", udp: true, cipher: chacha20-ietf-poly1305 }
    - { name: "🇺🇸 美国 01", type: "ss", server: "bppleman.com", port: 24050, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇺🇸 美国 02", type: "ss", server: "bppleman.com", port: 24051, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇺🇸 美国 03", type: "ss", server: "bppleman.com", port: 24052, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇺🇸 美国 04", type: "ss", server: "bppleman.com", port: 24053, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇺🇸 美国 05", type: "ss", server: "bppleman.com", port: 24054, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇨🇦 加拿大 01", type: "ss", server: "bppleman.com", port: 24060, password: "bppleman", udp: true, cipher: aes-128-gcm }
    - { name: "🇰🇷 韩国 01", type: "trojan", server: "bppleman.com", port: 60101, password: "bppleman", udp: true, sni: "video-caps.wetvinfo.com", skip-cert-verify: true }
    - { name: "🇬🇧 英国 01", type: "trojan", server: "bppleman.com", port: 60082, password: "bppleman", udp: true, sni: "video-caps.wetvinfo.com", skip-cert-verify: true }
    - { name: "🇫🇷 法国 01", type: "trojan", server: "bppleman.com", port: 60138, password: "bppleman", udp: true, sni: "video-caps.wetvinfo.com", skip-cert-verify: true }
    - { name: "🇩🇪 德国 01", type: "trojan", server: "bppleman.com", port: 60125, password: "bppleman", udp: true, sni: "video-caps.wetvinfo.com", skip-cert-verify: true }
    - { name: "🇦🇺 澳洲 01", type: "trojan", server: "bppleman.com", port: 60092, password: "bppleman", udp: true, sni: "video-caps.wetvinfo.com", skip-cert-verify: true }
proxy-groups:
    - { name: "BosLife", type: "select", proxies: [ 🇭🇰 香港, 🇨🇳🏳️‍🌈 台湾, 🇸🇬 新加坡, 🇯🇵 日本, 🇺🇸 美国, 🇨🇦 加拿大, 🇰🇷 韩国, 🇬🇧 英国, 🇫🇷 法国, 🇩🇪 德国, 🇦🇺 澳洲 ] }
    - { name: "Subscription Info", type: "select", proxies: [ 订阅提供商信息 ] }
    - { name: "🇭🇰 香港", type: "url-test", proxies: [ 🇭🇰 香港 01, 🇭🇰 香港 02, 🇭🇰 香港 03, 🇭🇰 香港 04, 🇭🇰 香港 05 - YouTube免广 ] }
    - { name: "🇨🇳🏳️‍🌈 台湾", type: "url-test", proxies: [ 🏳️‍🌈 台湾 01, 🏳️‍🌈 台湾 02, 🏳️‍🌈 台湾 03, 🏳️‍🌈 台湾 04, 🏳️‍🌈 台湾 05 ] }
    - { name: "🇸🇬 新加坡", type: "url-test", proxies: [ 🇸🇬 新加坡 01, 🇸🇬 新加坡 02, 🇸🇬 新加坡 03, 🇸🇬 新加坡 04, 🇸🇬 新加坡 05 ] }
    - { name: "🇯🇵 日本", type: "url-test", proxies: [ 🇯🇵 日本 01, 🇯🇵 日本 02, 🇯🇵 日本 03, 🇯🇵 日本 04, 🇯🇵 日本 05 ] }
    - { name: "🇺🇸 美国", type: "url-test", proxies: [ 🇺🇸 美国 01, 🇺🇸 美国 02, 🇺🇸 美国 03, 🇺🇸 美国 04, 🇺🇸 美国 05 ] }
    - { name: "🇨🇦 加拿大", type: "url-test", proxies: [ 🇨🇦 加拿大 01 ] }
    - { name: "🇰🇷 韩国", type: "url-test", proxies: [ 🇰🇷 韩国 01 ] }
    - { name: "🇬🇧 英国", type: "url-test", proxies: [ 🇬🇧 英国 01 ] }
    - { name: "🇫🇷 法国", type: "url-test", proxies: [ 🇫🇷 法国 01 ] }
    - { name: "🇩🇪 德国", type: "url-test", proxies: [ 🇩🇪 德国 01 ] }
    - { name: "🇦🇺 澳洲", type: "url-test", proxies: [ 🇦🇺 澳洲 01 ] }
rule-providers:
    Subscription_policy: { type: "http", url: "http://127.0.0.1/rule-provider/clash?interval=86400&policy[name]=DIRECT&policy[is_subscription]=true&sub_url=http://127.0.0.1:8080/subscription?token=bppleman", path: "./rule_providers/Subscription_policy.yaml", interval: 86400, size-limit: 0, format: "yaml", behavior: "classical" }
    BosLife_policy: { type: "http", url: "http://127.0.0.1/rule-provider/clash?interval=86400&policy[name]=BosLife&policy[is_subscription]=false&sub_url=http://127.0.0.1:8080/subscription?token=bppleman", path: "./rule_providers/BosLife_policy.yaml", interval: 86400, size-limit: 0, format: "yaml", behavior: "classical" }
    BosLife_no_resolve: { type: "http", url: "http://127.0.0.1/rule-provider/clash?interval=86400&policy[name]=BosLife&policy[option]=no-resolve&policy[is_subscription]=false&sub_url=http://127.0.0.1:8080/subscription?token=bppleman", path: "./rule_providers/BosLife_no_resolve.yaml", interval: 86400, size-limit: 0, format: "yaml", behavior: "classical" }
    BosLife_force_remote_dns: { type: "http", url: "http://127.0.0.1/rule-provider/clash?interval=86400&policy[name]=BosLife&policy[option]=force-remote-dns&policy[is_subscription]=false&sub_url=http://127.0.0.1:8080/subscription?token=bppleman", path: "./rule_providers/BosLife_force_remote_dns.yaml", interval: 86400, size-limit: 0, format: "yaml", behavior: "classical" }
    DIRECT_policy: { type: "http", url: "http://127.0.0.1/rule-provider/clash?interval=86400&policy[name]=DIRECT&policy[is_subscription]=false&sub_url=http://127.0.0.1:8080/subscription?token=bppleman", path: "./rule_providers/DIRECT_policy.yaml", interval: 86400, size-limit: 0, format: "yaml", behavior: "classical" }
    DIRECT_no_resolve: { type: "http", url: "http://127.0.0.1/rule-provider/clash?interval=86400&policy[name]=DIRECT&policy[option]=no-resolve&policy[is_subscription]=false&sub_url=http://127.0.0.1:8080/subscription?token=bppleman", path: "./rule_providers/DIRECT_no_resolve.yaml", interval: 86400, size-limit: 0, format: "yaml", behavior: "classical" }
    DIRECT_force_remote_dns: { type: "http", url: "http://127.0.0.1/rule-provider/clash?interval=86400&policy[name]=DIRECT&policy[option]=force-remote-dns&policy[is_subscription]=false&sub_url=http://127.0.0.1:8080/subscription?token=bppleman", path: "./rule_providers/DIRECT_force_remote_dns.yaml", interval: 86400, size-limit: 0, format: "yaml", behavior: "classical" }
rules:
    - RULE-SET,Subscription_policy,DIRECT
    - RULE-SET,BosLife_policy,BosLife
    - RULE-SET,BosLife_no_resolve,BosLife,no-resolve
    - RULE-SET,BosLife_force_remote_dns,BosLife,force-remote-dns
    - RULE-SET,DIRECT_policy,DIRECT
    - RULE-SET,DIRECT_no_resolve,DIRECT,no-resolve
    - RULE-SET,DIRECT_force_remote_dns,DIRECT,force-remote-dns
    - GEOIP,CN,DIRECT
    - MATCH,DIRECT
